test/select.jl

```diff
@@ -2346,6 +2365,300 @@ end
+        @test df2.b !== df2.c
+        df = DataFrame(a=1:3)
+        df2 = transform(df, sel => (x -> DataFrame(b=x, c=x, copycols=false)) => AsTable, copycols=false)
+        @test df.a === df2.a
+        @test df.a !== df2.b
+        @test df.a !== df2.c
+        @test df2.b !== df2.c
+
+        df = DataFrame(a=1:3)
+        a = df.a
+        select!(df, sel => (x -> (b=x, c=x)) => AsTable)
+        @test a === df.b
+        @test a !== df.c
+        df = DataFrame(a=1:3)
+        a = df.a
+        select!(df, sel => (x -> DataFrame(b=x, c=x, copycols=false)) => AsTable)
+        @test a === df.b
+        @test a !== df.c
+        df = DataFrame(a=1:3)
+        a = df.a
+        transform!(df, sel => (x -> (b=x, c=x)) => AsTable)
+        @test a === df.a
+        @test a !== df.b
+        @test a !== df.c
+        @test df.b !== df.c
+        df = DataFrame(a=1:3)
+        a = df.a
+        transform!(df, sel => (x -> DataFrame(b=x, c=x, copycols=false)) => AsTable)
+        @test a === df.a
+        @test a !== df.b
+        @test a !== df.c
+        @test df.b !== df.c
+    end
+
+    df = DataFrame(a=1:3)
+    df2 = select(df, :a, [:a, :a] => ((x, y) -> x) => :b, copycols=false)
+    @test df.a === df2.a
+    @test df.a == df2.b
+    @test df.a !== df2.b
+    df2 = transform(df, [:a, :a] => ((x, y) -> x) => :b, copycols=false)
+    @test df.a === df2.a
+    @test df.a == df2.b
+    @test df.a !== df2.b
+
+    df = DataFrame(a=1:3)
+    df2 = select(df, :a, [:a, :a] => ((x, y) -> (b=x, c=y)) => AsTable, copycols=false)
+    @test df.a === df2.a
+    @test df.a == df2.b
+    @test df.a !== df2.b
+    @test df.a == df2.c
+    @test df.a !== df2.c
+    df2 = transform(df, [:a, :a] => ((x, y) -> (b=x, c=y)) => AsTable, copycols=false)
+    @test df.a === df2.a
+    @test df.a == df2.b
+    @test df.a !== df2.b
+    @test df.a == df2.c
+    @test df.a !== df2.c
+    df2 = transform(df, [:a, :a] => ((x, y) -> (b=x, c=y)) => AsTable, :a => :d, copycols=false)
+    @test df.a === df2.a
+    @test df.a == df2.b
+    @test df.a !== df2.b
+    @test df.a == df2.c
+    @test df.a !== df2.c
+    @test df.a == df2.d
+    @test df.a !== df2.d
+
+    df = DataFrame(a=1:3, b=11:13, c=21:23)
+    df2 = select(df, [:a, :b, :c] => ((x, y, z) -> (c=x, d=y)) => AsTable, :, :c => :e, copycols=false)
+    @test df2.c === df.a
+    @test df2.d === df.b
+    @test df2.a !== df.a
+    @test df2.b !== df.b
+    @test df2.e === df.c
+
+    # mutlialias detection
+    df = DataFrame(a=1:3)
+    df.b = df.a
+    df2 = select(df, [:a, :b] => ((x, y) -> x) => :c, :a, :b, copycols=false)
+    @test df2.c === df.a === df.b
+    @test df2.a == df2.b == df.a
+    @test df2.a !== df.a
+    @test df2.b !== df.b
+
+    df = DataFrame(a=1:3)
+    df2 = transform(df, :a => lag)
+    @test df2.a_lag isa Vector
+    df2 = transform(df, :a => lag, copycols=false)
+    @test df2.a_lag isa ShiftedVector
+    transform!(df, :a => lag)
+    @test df.a_lag isa ShiftedVector
```

Comment:
I guess you refer to the line:
```
df2 = transform(df, :a => lag)
```
Copying is done because of this condition:
```
any(i -> vpar === parent(cdf[i]), parent_cols)
```
`transform` copies returnet columns ONLY if they would alias with source cols. And this is what I check. The test shows that since `lag` produces a view of a source column it gets materialized as a `Vector` to make sure `df2` and `df` do not share data. I think this is what we want. Do you agree? I will add a comment and update the docstring to clarify.

Reply:
Ah right of course we detect aliasing. I though the detection algorithm was less smart. But better fix the docstring to mention aliasing.

Reply:
I have updated it.